### PR TITLE
Transfer Incentives

### PR DESCRIPTION
**Pull request by michielderoos**
**Describe the Pull Request**

Adds org-level transfer incentives 

**Todo**

- [x] Link relevant [trello card](https://trello.com/b/PA2LhlOh/sempo-dev) or [related issue](https://github.com/teamsempo/SempoBlockchain/issues) to the pull request
- [x] Request review from relevant @person or team
- [x] QA your pull request. This includes running the app, login, testing changes etc.
- [x] Bump [backend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/config.py#L7) and/or [frontend](https://github.com/teamsempo/SempoBlockchain/blob/e3eb0f480e86d5a8ef2c1814127b70ff018671c4/app/package.json#L3) versions following Semver

### CODE REVIEW

File: app/server/me_api/credit_transfer.py [hunk 1]
```diff
@@ -282,8 +286,9 @@ def post(self):
                 'feedback': True,
             }
             return make_response(jsonify(response_object)), 400
-
+        incentive_transfers = []
         try:
+            # Question: Do we want the main transfer to go through if the insentive does not?
```
Comment (michielderoos) [inline app/server/me_api/credit_transfer.py:291]: Making a note on this question-- as it stands now hitting the incentive cap does not block the OG transaction from going through. It easily could though!

File: app/server/me_api/credit_transfer.py [hunk 2]
```diff
@@ -292,6 +297,21 @@ def post(self):
                                              transfer_use=transfer_use,
                                              transfer_mode=transfer_mode,
                                              uuid=uuid)
+            # Check the insentives to make sure that all of them pass, but don't actually make the payments yet (do_not_transact). 
+            # We want to make sure that all the transfers are valid before we allow them!
+            for incentive in g.active_organisation.incentives: # Do we want to use the active org, or the recipients org?
```
Comment (michielderoos) [inline app/server/me_api/credit_transfer.py:302]: Making a note on this question

File: app/server/models/credit_transfer.py [hunk 3]
```diff
@@ -296,7 +296,7 @@ def check_sender_transfer_limits(self):
         return relevant_transfer_limits
 
     def check_sender_has_sufficient_balance(self):
-        return self.sender_user and self.sender_transfer_account.balance - self.transfer_amount >= 0
```
Comment (michielderoos) [inline app/server/models/credit_transfer.py:299]: Why do these checks require a user _and_ sufficient balance?

File: app/server/utils/credit_transfer.py [hunk 4]
```diff
@@ -249,7 +249,7 @@ def make_payment_transfer(
                               transfer_mode=transfer_mode,
                               is_ghost_transfer=is_ghost_transfer)
 
-    make_cashout_incentive_transaction = False
+    make_cashout_incentive_transaction = False # is this obsolete now?
```
Comment (michielderoos) [inline app/server/utils/credit_transfer.py:252]: Making a note on this question

File: app/test_app/functional/me_api/test_me_credit_transfer.py [hunk 5]
```diff
@@ -16,6 +20,42 @@ def test_get_me_credit_transfer_api(test_client, create_credit_transfer, create_
     assert response.status_code == 201
     assert response.json['data']['credit_transfers'][0]['id'] is create_credit_transfer.id
 
+def test_valid_create_me_credit_transfer_api(
```
Comment (michielderoos) [inline app/test_app/functional/me_api/test_me_credit_transfer.py:23]: This test isn't really relevant to this PR, but it's something that really should be tested

**Comment by michielderoos** [comment]
[Add Tokenagent Incentive to peer-to-peer txn](https://trello.com/c/G6BX4ghf/590-add-tokenagent-incentive-to-peer-to-peer-txn)

File: app/server/models/credit_transfer.py [hunk 6]
```diff
@@ -296,7 +296,7 @@ def check_sender_transfer_limits(self):
         return relevant_transfer_limits
 
     def check_sender_has_sufficient_balance(self):
-        return self.sender_user and self.sender_transfer_account.unrounded_balance - Decimal(self.transfer_amount) >= 0
```
Comment (michielderoos) [inline app/server/models/credit_transfer.py:299]: Why does the balance check require a sender_user to exist?

File: app/server/models/credit_transfer.py (hunk 3 above)
Comment (enjeyw) [reply app/server/models/credit_transfer.py]: I was wondering the same thing so i removed lol!

File: app/server/models/incentive.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            - limit: Total spent by the sender user (BEFORE the currently evaluated incentive is applied)
+    """
+
+    __tablename__       = 'incentive'
+    incentive_rules = db.Column(JSON)
+    organisation_id = db.Column(db.Integer, db.ForeignKey(Organisation.id))
+
+    def handle_incentive(self, transfer, method, do_not_transact = False):
+        """
+        Handles insentives-- checks if they are applicable, then makes the insentive payment
```
Comment (enjeyw) [inline app/server/models/incentive.py:59]: nit: insentives --> incentives lol

File: app/server/models/incentive.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        if self.incentive_rules['incentive_type'] == 'FIXED':
+            incentive_amount = self.incentive_rules['incentive_amount']
+        elif self.incentive_rules['incentive_type'] == 'PERCENTAGE':
+            incentive_amount = transfer.transfer_amount * float(self.incentive_rules['incentive_amount'] / 100)
+
+        # Determine who gets the cash!
+        if self.incentive_rules['incentive_recipient'] == 'SENDER':
+            incentive_recipient_transfer_account = transfer.sender_transfer_account
+            incentive_recipient_user = transfer.sender_user
+        elif self.incentive_rules['incentive_recipient'] == 'RECIPIENT':
```
Comment (enjeyw) [inline app/server/models/incentive.py:75]: `elif self.incentive_rules['incentive_recipient'] == 'NICK'`:
🤑

File: app/server/models/incentive.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+            incentive_recipient_transfer_account = transfer.recipient_transfer_account
+            incentive_recipient_user = transfer.recipient_user
+
+        # Process restrictions, apply them to sender_user
+        for r in self.incentive_rules['restrictions']:
```
Comment (enjeyw) [inline app/server/models/incentive.py:80]: could it make more sense to integrate these into the limit code to bring all the limits into one place?

File: app/server/utils/credit_transfer.py (hunk 4 above)
Comment (enjeyw) [reply app/server/utils/credit_transfer.py:254]: yep it ded

File: app/server/me_api/credit_transfer.py (hunk 2 above)
Comment (enjeyw) [reply app/server/me_api/credit_transfer.py:302]: goooood question. i mean they're generally one and the same thing, but probably active org?

File: app/server/me_api/credit_transfer.py [hunk 10]
```diff
@@ -102,6 +103,8 @@ def post(self):
 
         is_sending = post_data.get('is_sending', False)
 
+        method = 'ANY'
```
Comment (enjeyw) [inline app/server/me_api/credit_transfer.py:106]: we already have a 'transfer_mode' on credit transfers (which is super neglected and under used) - but i think this should use that...

**Review by enjeyw — CHANGES_REQUESTED** [review]
clean overall, few small suggestions/thoughts!

File: app/server/models/incentive.py (hunk 7 above)
Comment (michielderoos) [reply app/server/models/incentive.py:59]: 🙃

File: app/server/models/incentive.py (hunk 9 above)
Comment (michielderoos) [reply app/server/models/incentive.py:81]: I think it depends on what you think we should do about [this](https://github.com/teamsempo/SempoBlockchain/pull/479#discussion_r497099036). To me, I think it makes sense to allow transactions to continue after the incentives have been exhausted. If we do block them though, I can see integrating them into limits. Does that make sense?

**Comment by tristanhcole** [comment]
Outdated

**Pull request closed without merging**